### ISSUE TITLE
TOML parser produces incorrect output when an array is followed by an inline comment

### ISSUE THREAD

**Issue by Jakob-Strobl** [denoland/deno#7072]
## Brief
When an array is followed by a comment, the TOML parser produces incorrect output. This issue is quite weird, but I tracked down a few cases of where the parser produces invalid output.

This issue and the examples below are verbose, but provide a clearer picture of the issue(s) with the parser. 

**Case 1** seems to be the **root cause**. 
**Case 2** and **Case 3** provide examples of other weird behaviors derived from **Case 1**.

**Deno Versions:**
deno 1.3.0
v8 8.6.334
typescript 3.9.7
deno-std 0.65.0

### Valid Case (output is correct)
```ts
console.log(parse(`
[deno] # this comment is fine
is_not_node = ["that", "is", "true"]
this_string_is_skipped = "no"
this_is_also_skipped = false 

[toml] 
toml_merged = false
toml_string_merged = "no"
`));
```

**Correct output:** This will be used as a reference for the following cases.
```ts
{
  deno: {
    is_not_node: [ "that", "is", "true" ],
    this_string_is_skipped: "no",
    this_is_also_skipped: false
  },
  toml: { toml_merged: false, toml_string_merged: "no" }
}
```
---
### Case 1
A comment after an array causes two problems: 
1. ```[toml]```'s key-values to merge into ```[deno]```.
2. The key-values in ```[deno]``` after the comment disappear

```ts
console.log(parse(`
[deno] # this comment is fine
is_not_node = ["that", "is", "true"] # This comment causes Case 1
this_string_is_skipped = "yes"
this_is_also_skipped = true 

[toml] 
toml_merged = true
toml_string_merged = "yes"
`))
```

**Ouput:**
```ts
{
  deno: {
    is_not_node: [ "that", "is", "true" ],
    toml_merged: true,
    toml_string_merged: "yes"
  }
}
```
**Expected output:** The output should be similar to **Valid Case**'s output.

---
### Case 2

When there is a comment after an array **and** the next table has a comment inline after it, the parser produces the wrong output:
1. Only the first table, ```[deno]```, is in the output
2. All the key-values of ```[deno]``` disappear
3. The second table, ```[toml]```, and its key-values disappear

```ts 
console.log(parse(`
[deno] # this comment is fine
is_not_node = ["that", "is", "true"] # This comment causes Case 1
this_string_is_skipped = "yes"
this_is_also_skipped = true 

[toml] # This comment in tandem with Case 1's comment, causes Case 2
toml_merged = true
toml_string_merged = "idk it is all gone"
`))
```
**Output:**
```ts
{ deno: {} }
```
**Expected output:** The output should be similar to **Valid Case**'s output.

**Note:**
- It seems to be if and only if the following table, ```[toml]```, has a comment after it 
- If there are more tables and each table is also followed by an inline comment, they also disappear

---
### Case 3
When there is a comment after an array, and then one of the next key-values is also an array the output changes:
1. An array following the commented array will disappear. In this case its the array called ```this_array_is_skipped```
2. Any key-values in between the commented array and the second array will also disappear 
3. The values after the second array, ```this_array_is_skipped```, will **NOT** disappear and are shown correctly in the output
4. The second table, ```[toml]```, is now parsed correctly even though the table definition has an inline comment

```ts
console.log(parse(`
[deno] # this comment is fine
is_not_node = ["that", "is", "true"]  # This comment causes Case 1
this_string_is_skipped = "yes"
this_array_is_skipped = ["yes, this array is skipped.", "The following key-values are not skipped."]
this_is_also_skipped = false 

[toml] # This comment used to cause an issue, but its fine now due to the second array
toml_merged = false
toml_string_merged = "no"
`))
```

**Output:**
```ts
{
  deno: { is_not_node: [ "that", "is", "true" ], this_is_also_skipped: false },
  toml: { toml_merged: false, toml_string_merged: "no" }
}
````

**Expected output:** The output should be similar to **Valid Case**'s output.

Note: 
- The second array, ```this_array_is_skipped```, seems to reset the parser. Any TOML content following the second array is parsed correctly.
- If the second array is followed by a comment, the output is the same as **Case 2**

**Comment by Jakob-Strobl** [comment]
I'll take stab at this issue, but won't be able to start for a few days. I'll update if I meet any roadblocks.

**Comment by Jakob-Strobl** [comment]
I finally found the time to start working on this issue today. I was able to track down the source of the bug in the TOML parser. 

Unfortunately this issue seems to span multiple functions/dependencies, so it's not a quick 1-2 line fix.
Instead, I thought it would be a good idea to strip all comments during the _sanitize() phase of the parser. 
By stripping all the comments before actually parsing the TOML lines, we can simplify the regular expressions and parsing functions for the various JSON data types. 

I got a working solution that passes all the TOML test cases, but I need to clean it up and broaden test coverage first.
